src/pudl/output/sec10k.py

```diff
@@ -106,3 +110,135 @@ def out_sec10k__changelog_company_name(
+    # parent companies in the ownership table.
+    parent_co_merge_key = [
+        "filename_sec10k",
+        "parent_company_central_index_key",
+        "report_date",
+    ]
+    n_parent_dupes = _get_n_duplicate_filer_records(
+        out_sec10k__quarterly_company_information.rename(
+            columns={"central_index_key": "parent_company_central_index_key"}
+        ),
+        core_sec10k__quarterly_exhibit_21_company_ownership,
+        key=parent_co_merge_key,
+    )
+    if n_parent_dupes > 0:
+        logger.warning(
+            f"""There are {n_parent_dupes} duplicate filenames, CIKs, and report date records that are parent companies in the Ex. 21 ownership table."""
+        )
+    # merge parent attributes on
+    parents_info_df = out_sec10k__quarterly_company_information.add_prefix(
+        "parent_company_"
+    ).drop(columns=["parent_company_company_name"])
+    df = core_sec10k__quarterly_exhibit_21_company_ownership.merge(
+        parents_info_df,
+        how="left",
+        left_on=["filename_sec10k", "parent_company_central_index_key", "report_date"],
+        right_on=[
+            "parent_company_filename_sec10k",
+            "parent_company_central_index_key",
+            "parent_company_report_date",
+        ],
+    ).drop(columns=["parent_company_report_date", "parent_company_filename_sec10k"])
+
+    # merge central index key on for subsidiaries that file a 10k
+    df = df.merge(
+        core_sec10k__assn_exhibit_21_subsidiaries_and_filers,
+        how="left",
+        on="subsidiary_company_id_sec10k",
+    ).rename(columns={"central_index_key": "subsidiary_company_central_index_key"})
+
+    # merge utility_id_eia onto subsidiaries
+    df = df.merge(
+        core_sec10k__assn_exhibit_21_subsidiaries_and_eia_utilities,
+        how="left",
+        on="subsidiary_company_id_sec10k",
+    )
+
+    # merge utility name onto subsidiaries
+    df = df.merge(core_eia__entity_utilities, how="left", on="utility_id_eia").rename(
+        columns={
+            "utility_id_eia": "sub_only_utility_id_eia",
+            "utility_name_eia": "sub_only_utility_name_eia",
+        }
+    )
+    # Log the number of non unique filename, CIK, report_date records from the
+    # information table that appear as subsidiary companies in the ownership table.
+    sub_co_merge_key = [
+        "filename_sec10k",
+        "subsidiary_company_central_index_key",
+        "report_date",
+    ]
+    n_sub_dupes = _get_n_duplicate_filer_records(
+        out_sec10k__quarterly_company_information.rename(
+            columns={"central_index_key": "subsidiary_company_central_index_key"}
+        ),
+        df,
+        key=sub_co_merge_key,
+    )
+    if n_sub_dupes > 0:
+        logger.warning(
+            f"""There are {n_sub_dupes} duplicate filenames, CIKs, and report date records that are subsidiary companies in the Ex. 21 ownership table."""
+        )
```

Comment:
So we warn, but then do the merge anyway, which I think means we're broadcasting and creating duplicate records? One for each version of the company info associated with that CIK and report date?  Is that right? Or does the inclusion of the filename prevent that an end up forcing it to be a 1-to-1 merge?

Reply:
Why do we think that the first one is the right one to keep when there are multiple matches?